plugins/rebalance/rest.go

```diff
@@ -0,0 +1,276 @@
+	rebalanceapi "github.com/gluster/glusterd2/plugins/rebalance/api"
+	"github.com/gorilla/mux"
+	"github.com/pborman/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+func createRebalanceInfo(rebalance *rebalanceapi.RebalanceInfo) (*rebalanceapi.RebalanceInfo, error) {
+	v := new(rebalanceapi.RebalanceInfo)
+	v.Volname = rebalance.Volname
+	v.RebalanceID = uuid.NewRandom()
+	v.Status = rebalance.Status
+	v.Cmd = rebalance.Cmd
+	v.CommitHash = rebalance.CommitHash
+	return v, nil
+}
+
+func rebalanceStart(w http.ResponseWriter, r *http.Request) {
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	// collect inputs from url
+	volname := mux.Vars(r)["volname"]
+	rebalance := new(rebalanceapi.RebalanceInfo)
+	req := new(rebalanceapi.RebalanceStartReq)
+	if err := restutils.UnmarshalRequest(r, &req); err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusUnprocessableEntity, "Unable to parse the request", api.ErrCodeDefault)
+		return
+	}
+
+	vol, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(r.Context(), w, http.StatusNotFound, "Invalid volume", api.ErrCodeDefault)
+		return
+	}
+
+	rebalinfo, err := GetRebalanceInfo(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, err.Error(), api.ErrCodeDefault)
+		return
+	}
+
+	if vol.State != volume.VolStarted {
+		restutils.SendHTTPError(r.Context(), w, http.StatusBadRequest, "Volume not started", api.ErrCodeDefault)
+		return
+	}
+
+	if vol.DistCount == 1 {
+		restutils.SendHTTPError(r.Context(), w, http.StatusBadRequest, "Volume is not distributed volume or contain only 1 brick",
+			api.ErrCodeDefault)
+		return
+	}
+
+	// Check rebalance is already in process
+	if rebalinfo.Status == rebalanceapi.GfDefragStatusStarted {
+		restutils.SendHTTPError(r.Context(), w, http.StatusBadRequest, "Rebalance is already in progess",
+			api.ErrCodeDefault)
+		return
+	}
+
+	// Check for remove- brick pending
+	//TODO
+
+	//A simple transaction to start rebalance
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+	lock, unlock, err := transaction.CreateLockSteps(volname)
+	if err != nil {
+		restutils.SendHTTPError(r.Context(), w, http.StatusInternalServerError, err.Error(), api.ErrCodeDefault)
+		return
+	}
+
+	txn.Nodes = vol.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc: "rebal-start.Commit",
+			Nodes:  txn.Nodes,
+		},
+		{
+			DoFunc: "rebal-start.StoreVolume",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
+		},
+		unlock,
+	}
+
+	err = txn.Ctx.Set("volname", volname)
+	if err != nil {
+		logger.WithError(err).Error("failed to set request in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err.Error(), api.ErrCodeDefault)
+		return
+	}
+
+	rebalance.Volname = volname
+	rebalance.Status = rebalanceapi.GfDefragStatusStarted
+	rebalance.Cmd = rebalanceapi.GfDefragCmdStart
+	setcommithash(rebalance)
+	rebal, err := createRebalanceInfo(rebalance)
```

Comment:
You can change this function to accept Volume Name and return Rebalance Info. So that you can avoid `rebalance` variable.

Reply:
I tried it with volname but it actually didn't work as expected.